File: server/models/User.model.js
import mongoose from 'mongoose';
import bcrypt from 'bcryptjs';

const userSchema = new mongoose.Schema({
  firstName: { type: String, required: true, trim: true },
  lastName: { type: String, required: true, trim: true },
  email: { type: String, required: true, unique: true, lowercase: true, trim: true },
  phone: { type: String, required: true },
  password: { type: String, required: true, minlength: 8 },
  country: { type: String, required: true },
  currency: { type: String, default: 'USD' },
  role: { type: String, enum: ['user', 'admin', 'compliance', 'support'], default: 'user' },
  
  // Account status
  isEmailVerified: { type: Boolean, default: false },
  isPhoneVerified: { type: Boolean, default: false },
  isActive: { type: Boolean, default: true },
  isSuspended: { type: Boolean, default: false },
  suspendedReason: { type: String },
  
  // KYC
  kycStatus: { type: String, enum: ['pending', 'submitted', 'approved', 'rejected'], default: 'pending' },
  kycRejectedReason: { type: String },
  
  // Verification tokens
  emailVerificationToken: { type: String },
  emailVerificationExpires: { type: Date },
  passwordResetToken: { type: String },
  passwordResetExpires: { type: Date },
  otpCode: { type: String },
  otpExpires: { type: Date },
  
  // 2FA
  twoFactorEnabled: { type: Boolean, default: false },
  twoFactorSecret: { type: String },
  
  // Profile
  avatar:          { type: String }, // Cloudinary URL
  avatarPublicId:  { type: String }, // Cloudinary publicId for deletion
  dateOfBirth: { type: Date },
  address: {
    street: String,
    city: String,
    state: String,
    postalCode: String,
    country: String,
  },
  
  // Stats
  totalTransfers: { type: Number, default: 0 },
  totalAmountSent: { type: Number, default: 0 },
  lastLogin: { type: Date },
  loginAttempts: { type: Number, default: 0 },
  lockUntil: { type: Date },
}, { timestamps: true });

// Hash password before saving
userSchema.pre('save', async function(next) {
  if (!this.isModified('password')) return next();
  this.password = await bcrypt.hash(this.password, 12);
  next();
});

// Compare password method
userSchema.methods.comparePassword = async function(candidatePassword) {
  return await bcrypt.compare(candidatePassword, this.password);
};

// Virtual: full name
userSchema.virtual('fullName').get(function() {
  return `${this.firstName} ${this.lastName}`;
});

userSchema.set('toJSON', { virtuals: true });

export default mongoose.model('User', userSchema);

File: transfastApp/lib/features/auth/models/user_model.dart
// lib/features/auth/models/user_model.dart
import 'dart:convert';

class UserModel {
  final String id;
  final String firstName;
  final String lastName;
  final String email;
  final String phone;
  final String country;
  final String role;
  final String kycStatus;
  final bool isEmailVerified;
  final bool isActive;
  final bool isSuspended;
  final int totalTransfers;
  final double totalAmountSent;
  final String? lastLogin;
  final String currency;
  final String? avatar;        // Cloudinary URL
  final String? avatarPublicId;

  const UserModel({
    required this.id,
    required this.firstName,
    required this.lastName,
    required this.email,
    required this.phone,
    required this.country,
    required this.role,
    required this.kycStatus,
    required this.isEmailVerified,
    required this.isActive,
    required this.isSuspended,
    required this.totalTransfers,
    required this.totalAmountSent,
    this.lastLogin,
    this.currency = 'USD',
    this.avatar,
    this.avatarPublicId,
  });

  String get fullName  => '$firstName $lastName';
  String get initials  => '${firstName[0]}${lastName[0]}'.toUpperCase();
  bool get isAdmin     => ['admin', 'compliance', 'support'].contains(role);
  bool get isKycApproved => kycStatus == 'approved';

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id:              json['_id'] ?? json['id'] ?? '',
      firstName:       json['firstName'] ?? '',
      lastName:        json['lastName']  ?? '',
      email:           json['email']     ?? '',
      phone:           json['phone']     ?? '',
      country:         json['country']   ?? '',
      role:            json['role']      ?? 'user',
      kycStatus:       json['kycStatus'] ?? 'pending',
      isEmailVerified: json['isEmailVerified'] ?? false,
      isActive:        json['isActive']        ?? true,
      isSuspended:     json['isSuspended']     ?? false,
      totalTransfers:  json['totalTransfers']  ?? 0,
      totalAmountSent: (json['totalAmountSent'] ?? 0).toDouble(),
      lastLogin:       json['lastLogin'],
      currency:        json['currency'] ?? 'USD',
      avatar:          json['avatar'],
      avatarPublicId:  json['avatarPublicId'],
    );
  }

  Map<String, dynamic> toJson() => {
    '_id':            id,
    'firstName':      firstName,
    'lastName':       lastName,
    'email':          email,
    'phone':          phone,
    'country':        country,
    'role':           role,
    'kycStatus':      kycStatus,
    'isEmailVerified':isEmailVerified,
    'isActive':       isActive,
    'isSuspended':    isSuspended,
    'totalTransfers': totalTransfers,
    'totalAmountSent':totalAmountSent,
    'lastLogin':      lastLogin,
    'currency':       currency,
    if (avatar != null)         'avatar':        avatar,
    if (avatarPublicId != null) 'avatarPublicId':avatarPublicId,
  };

  String toJsonString() => json.encode(toJson());
  factory UserModel.fromJsonString(String s) => UserModel.fromJson(json.decode(s));

  UserModel copyWith({
    String? firstName, String? lastName, String? phone,
    String? kycStatus, bool? isEmailVerified,
    int? totalTransfers, double? totalAmountSent,
    String? avatar, String? avatarPublicId,
    bool clearAvatar = false,
  }) {
    return UserModel(
      id: id, email: email, country: country, role: role,
      isActive: isActive, isSuspended: isSuspended, lastLogin: lastLogin, currency: currency,
      firstName:       firstName       ?? this.firstName,
      lastName:        lastName        ?? this.lastName,
      phone:           phone           ?? this.phone,
      kycStatus:       kycStatus       ?? this.kycStatus,
      isEmailVerified: isEmailVerified ?? this.isEmailVerified,
      totalTransfers:  totalTransfers  ?? this.totalTransfers,
      totalAmountSent: totalAmountSent ?? this.totalAmountSent,
      avatar:          clearAvatar ? null : (avatar ?? this.avatar),
      avatarPublicId:  clearAvatar ? null : (avatarPublicId ?? this.avatarPublicId),
    );
  }
}
